src/snowflake/snowpark/ml/transformer.py

```diff
@@ -0,0 +1,395 @@
+#
+# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
+#
+
+
+from typing import List, Optional, Tuple, Union
+
+from snowflake.snowpark._internal.utils import (
+    TempObjectType,
+    random_name_for_temp_object,
+)
+from snowflake.snowpark.dataframe import DataFrame
+from snowflake.snowpark.functions import (
+    array_agg,
+    array_cat,
+    array_construct,
+    array_construct_compact,
+    array_insert,
+    array_slice,
+    col,
+    iff,
+    lit,
+)
+from snowflake.snowpark.ml.utils import encoder_fit, scaler_fit
+
+
+class Transformer:
+    def __init__(
+        self,
+        input_cols: Optional[List[str]] = None,
+        output_cols: Optional[List[str]] = None,
+    ):
+        self._input_cols = input_cols
+        self._output_cols = output_cols
+
+    @property
+    def input_cols(self) -> List[str]:
+        return self._input_cols
+
+    @input_cols.setter
+    def input_cols(self, value: List[str]) -> None:
+        self._input_cols = value
+
+    @property
+    def output_cols(self) -> List[str]:
+        return self._output_cols
+
+    @output_cols.setter
+    def output_cols(self, value: List[str]) -> None:
+        self._output_cols = value
+
+    def fit(self, df: DataFrame) -> "Transformer":
+        pass
+
+    def transform(self, df: DataFrame) -> DataFrame:
+        pass
+
+
+class StandardScaler(Transformer):
+    def __init__(
+        self,
+        input_cols: Optional[List[str]] = None,
+        output_cols: Optional[List[str]] = None,
+    ):
+        super().__init__(input_cols, output_cols)
+        self._states_table_cols = (
+            [f"states_{c}" for c in self._input_cols] if self._input_cols else None
+        )
+        self._states_table_name = random_name_for_temp_object(TempObjectType.TABLE)
+
+    @Transformer.input_cols.setter
+    def input_cols(self, value: List[str]) -> None:
+        self._input_cols = value
+        self._states_table_cols = (
+            [f"states_{c}" for c in self._input_cols] if self._input_cols else None
+        )
+
+    def fit(self, df: DataFrame) -> "StandardScaler":
+        if not self._input_cols:
+            raise ValueError("Input columns can not be None")
+
+        scaler_fit(self, df, ["mean", "stddev_pop"])
+        return self
+
+    def transform(self, df: DataFrame) -> DataFrame:
+        """
+        The standard score of a sample x is calculated as:
+        z = (x - mean) / stddev
+        """
+        if not self._output_cols or len(self._input_cols) != len(self._output_cols):
+            raise ValueError(
+                "The number of input column and output column does not match or output column is empty"
+            )
+        states_table = df._session.table(self._states_table_name)
+        return (
```

Comment:
My bad, actually we should also retain the original columns for scaler. Could you try to use with_columns to add these output cols, instead of using to_df?